websockify/websocket.py

```diff
@@ -1,1109 +1,825 @@
+                self._received_close = True
+
+                if self._sent_close:
+                    self._close()
+                    return ''.encode("ascii")
+
+                if not frame["fin"]:
+                    self.shutdown(socket.SHUT_RDWR, 1003, "Unsupported: Fragmented close")
+                    continue
+
+                code = None
+                reason = None
+                if len(frame["payload"]) >= 2:
+                    code = struct.unpack(">H", buf[:2])
+                    if len(frame["payload"]) > 2:
+                        reason = frame["payload"][2:]
+                        try:
+                            reason = reason.decode("UTF-8")
+                        except UnicodeDecodeError:
+                            self.shutdown(socket.SHUT_RDWR, 1002, "Procotol error: Invalid UTF-8 in close")
+                            continue
+
+                if code is None:
+                    self.close_code = 1005
+                    self.close_reason = "No close status code specified by peer"
+                else:
+                    self.close_code = code
+                    if reason is not None:
+                        self.close_reason = reason
+
+                self.shutdown(code, reason)
+                return ''.encode("ascii")
+            elif frame["opcode"] == 0x9:
+                if not frame["fin"]:
+                    self.shutdown(socket.SHUT_RDWR, 1003, "Unsupported: Fragmented ping")
+                    continue
+
+                self.handle_ping(frame["payload"])
+            elif frame["opcode"] == 0xA:
+                if not frame["fin"]:
+                    self.shutdown(socket.SHUT_RDWR, 1003, "Unsupported: Fragmented pong")
+                    continue
+
+                self.handle_pong(frame["payload"])
+            else:
+                self.shutdown(socket.SHUT_RDWR, 1003, "Unsupported: Unknown opcode 0x%02x" % frame["opcode"])
+
+        return None
+
+    def _flush(self):
+        # Writes pending data to the socket
+        if not self._send_buffer:
+            return
+
+        assert self.socket is not None
+
+        try:
+            sent = self.socket.send(self._send_buffer)
+        except (socket.error, OSError):
+            exc = sys.exc_info()[1]
+            if hasattr(exc, 'errno'):
+                err = exc.errno
+            else:
+                err = exc[0]
+
+            if err == errno.EWOULDBLOCK:
+                raise WebSocketWantWriteError
+
+            raise
+
+        self._send_buffer = self._send_buffer[sent:]
+
+        if self._send_buffer:
+            raise WebSocketWantWriteError
+
+        # We had a pending close and we've flushed the buffer,
+        # time to end things
+        if self._received_close and self._sent_close:
+            self._close()
+
+    def _send(self, data):
+        # Queues data and attempts to send it
+        self._send_buffer += data
+        self._flush()
+
+    def _send_str(self, string):
```

Comment:
perhaps name it something besides `_send_str` if the doc string says it doesn't send ;-)

(`_queue_str`, for instance?)